File: codbex-number-generator/service/generator.ts
import { NumberRepository } from "../gen/codbex-number-generator/data/Settings/NumberRepository";

export class NumberGeneratorService {

    private readonly numberRepository = new NumberRepository();

    public generate(id: number): string | undefined {
        try {
            const entity = this.numberRepository.findById(id);

            if (!entity) {
                throw new Error(`Entity with id ${id} not found`);
            }

            entity.Value! += 1;
            this.numberRepository.update(entity);

            const zeroPad = (num: number, places: number) => String(num).padStart(places, '0');
            const result: string = (entity.Prefix ? entity.Prefix : "")
                + zeroPad(entity.Value!, entity.Length!
                    - (entity.Prefix ? entity.Prefix.length : 0));

            return result;

        } catch (error: any) {
            console.error(error);
        }
    }

    public generateByType(genType: string): string | undefined {
        try {

            const entity = this.numberRepository.findAll()
                .filter(item => item.Type === genType).at(0);

            if (!entity) {
                throw new Error(`Entity with type ${genType} not found`);
            }

            entity.Value! += 1;
            this.numberRepository.update(entity);

            const zeroPad = (num: number, places: number) => String(num).padStart(places, '0');
            const result: string = (entity.Prefix ? entity.Prefix : "")
                + zeroPad(entity.Value!, entity.Length!
                    - (entity.Prefix ? entity.Prefix.length : 0));

            return result;

        } catch (error: any) {
            console.error(error);
        }
    }

}

File: codbex-number-generator/service/Generator.java
package service;

import gen.codbex_number_generator.data.settings.NumberEntity;
import gen.codbex_number_generator.data.settings.NumberRepository;

import org.eclipse.dirigible.engine.java.annotations.Inject;

import java.util.List;
import java.util.Map;

public class Generator {

    @Inject
    private NumberRepository numberRepository;

    public String generate(Integer id) {
        try {
            NumberEntity entity = numberRepository.findOne(id)
                .orElseThrow(() ->
                    new RuntimeException("Entity with id " + id + " not found"));

            return generateNumber(entity);

        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public String generateByType(String genType) {
        try {
            List < NumberEntity > entities = numberRepository.query(
                "from NumberEntity e where e.Type = :type",
                Map.of("type", genType)
            );

            if (entities.isEmpty()) {
                throw new RuntimeException(
                    "Entity with type " + genType + " not found");
            }

            return generateNumber(entities.get(0));

        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private String generateNumber(NumberEntity entity) {
        entity.Value = entity.Value + 1;
        numberRepository.update(entity);

        String prefix = entity.Prefix != null ? entity.Prefix : "";

        int totalLength = entity.Length != null ? entity.Length : 0;
        int numericLength = Math.max(0, totalLength - prefix.length());

        String paddedValue = String.format(
            "%0" + numericLength + "d",
            entity.Value
        );

        return prefix + paddedValue;
    }
}